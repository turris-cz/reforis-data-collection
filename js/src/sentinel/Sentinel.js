/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import { ForisForm, useAPIGet } from "foris";

import API_URLs from "../API";
import SentinelState from "./SentinelState/SentinelState";
import SentinelOptions from "./SentinelOptions/SentinelOptions";
import DisabledIfNotAccepted from "../utils/DisabledIfNotAccepted";

export default function Sentinel() {
    // We had to move API call up to pass it to the `postCallback` function
    const [sentinelComponentsState, getSentinelComponentsState] = useAPIGet(
        API_URLs.state
    );

    useEffect(() => {
        getSentinelComponentsState();
    }, [getSentinelComponentsState]);

    return (
        <>
            <h1>{_("Sentinel")}</h1>
            <p>
                {_(
                    "Sentinel is a Turris threat detection and attack prevention system, which provides dynamic firewall and statistics. Here you can set up several Sentinel components which take part in the threat detection subsystem."
                )}
            </p>
            <div id="sentinel-state" />
            <ForisForm
                forisConfig={{
                    endpoint: API_URLs.settings,
                }}
                postCallback={getSentinelComponentsState}
                prepDataToSubmit={prepDataToSubmit}
                validator={validator}
            >
                <SentinelState
                    apiState={sentinelComponentsState}
                    states={sentinelComponentsState}
                />
                <DisabledIfNotAccepted>
                    <SentinelOptions />
                </DisabledIfNotAccepted>
            </ForisForm>
        </>
    );
}

function prepDataToSubmit(data) {
    data.eula = parseInt(data.eula);
    if (!data.token) delete data.token;
    delete data.modules.survey.installed;
    delete data.modules.fwlogs.installed;
    delete data.modules.minipot.installed;
    return data;
}

// Hack to disable submit button
function validator(formData) {
    if (formData.eula !== 1) return { eula: 2 };
    return null;
}
