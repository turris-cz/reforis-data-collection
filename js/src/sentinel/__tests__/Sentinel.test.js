/*
 * Copyright (C) 2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    wait,
    fireEvent,
} from "foris/testUtils/customTestRender";

import Sentinel from "../Sentinel";
import mockAxios from "jest-mock-axios";
import { sentinelStateFixture, sentinelSettingsFixture } from "./__fixtures__/sentinelFixture";

describe("<Sentinel />", () => {
    let asFragment;
    let firstRender;
    let getByText;

    beforeEach(async () => {
        const renderRes = render(<Sentinel />);
        mockAxios.mockResponse({ data: sentinelSettingsFixture });
        getByText = renderRes.getByText;
        asFragment = renderRes.asFragment;

        await wait(() => getByText("Enable Firewall Logs"));

        mockAxios.mockResponse({ data: sentinelStateFixture });

        await wait(() => getByText("Sentinel Proxy"));
        firstRender = renderRes.asFragment();
    });

    it("Snapshot of Sentinel", () => {
        getByText("Sentinel Components");
        expect(firstRender).toMatchSnapshot();
    });
    
    it("Success post request", () => {
        fireEvent.click(getByText("Enable Firewall Logs"));
    
        fireEvent.click(getByText("Save"));
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/data-collection/api/settings",
            {
                eula: 1,
                modules: {
                    minipot: {
                        enabled: false,
                        protocols: {
                            ftp: true,
                            http: true,
                            smtp: true,
                            telnet: true,
                        },
                    },
                    nikola: { enabled: true },
                    survey: { enabled: false },
                },
                token: "random_token",
            },
            expect.anything()
        );
    });
});