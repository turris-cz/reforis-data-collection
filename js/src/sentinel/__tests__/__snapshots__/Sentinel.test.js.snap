// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sentinel /> Snapshot of Sentinel 1`] = `
<DocumentFragment>
  <h1>
    Sentinel
  </h1>
  <p>
    Sentinel is a Turris threat detection and attack prevention system, which provides dynamic firewall and statistics. Here you can set up several Sentinel components which take part in the threat detection subsystem.
  </p>
  <div
    id="sentinel-state"
  >
    <div
      class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
    >
      <h2>
        Sentinel State
      </h2>
      <p>
        To view Sentinel data, please visit 
        <a
          href="https://view.sentinel.turris.cz"
          rel="noopener noreferer"
          target="_blank"
        >
          Sentinel View
          <sup>
            <i
              class="fas fa-external-link-alt fa-xs ml-1"
            />
          </sup>
        </a>
        . There you can also see and filter 
        <a
          href="https://view.sentinel.turris.cz/api/device/add?token=random_token"
          rel="noopener noreferer"
          target="_blank"
        >
          data specific for your device
          <sup>
            <i
              class="fas fa-external-link-alt fa-xs ml-1"
            />
          </sup>
        </a>
        .
      </p>
      <div
        class="row no-gutters justify-content-center"
      >
        <div
          class="col-md-6 d-flex flex-column align-items-center justify-content-center mt-3 mt-md-0"
        >
          <span
            class="text-success"
          >
            <i
              class="fas fa-database fa-5x mb-1"
              title="Status: Running"
            />
          </span>
          <p
            class="h4 mt-3"
          >
            Sentinel
          </p>
        </div>
        <div
          class="col-md-6"
        >
          <div
            class="card-body"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table table-borderless table-hover mb-0"
              >
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      Sentinel Proxy
                    </th>
                    <td>
                      <span
                        class="text-success"
                      >
                        <i
                          class="fas fa-check"
                        />
                      </span>
                    </td>
                    <td
                      class="text-center"
                    >
                      <span
                        class="badge badge-success"
                        title="Status: Running"
                      >
                        Running
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Firewall Logs
                    </th>
                    <td>
                      <span
                        class="text-warning"
                      >
                        <i
                          class="fas fa-exclamation-triangle"
                        />
                      </span>
                    </td>
                    <td
                      class="text-center"
                    >
                      <span
                        class="badge badge-warning"
                        title="Status: Disabled"
                      >
                        Disabled
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Minipots
                    </th>
                    <td>
                      <span
                        class="text-warning"
                      >
                        <i
                          class="fas fa-exclamation-triangle"
                        />
                      </span>
                    </td>
                    <td
                      class="text-center"
                    >
                      <span
                        class="badge badge-warning"
                        title="Status: Disabled"
                      >
                        Disabled
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Survey
                    </th>
                    <td>
                      <span
                        class="text-warning"
                      >
                        <i
                          class="fas fa-exclamation-triangle"
                        />
                      </span>
                    </td>
                    <td
                      class="text-center"
                    >
                      <span
                        class="badge badge-warning"
                        title="Status: Disabled"
                      >
                        Disabled
                      </span>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          for="4"
        >
          Device token
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="4"
            readonly=""
            type="text"
            value="random_token"
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-outline-secondary"
              type="button"
            >
              <span>
                Copy
              </span>
            </button>
          </div>
        </div>
        <small
          class="form-text text-muted"
        >
          That token allows you to access statistics from your device on Sentinel View.
        </small>
      </div>
    </div>
  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <form>
      <h2>
        Sentinel Components
      </h2>
      <p>
        You can select specific components that you want to enable or disable.
      </p>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-switch"
        >
          <input
            class="custom-control-input"
            id="1"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="1"
          >
            Enable Firewall Logs
          </label>
          <small
            class="form-text text-muted mt-0 mb-3"
          >
            Firewall Logs are logs gathered from iptables firewall. If enabled, Sentinel uses them to monitor packets coming from outside network and trying to connect to potentially vulnerable local services. These techniques, also known as "port scans" usually try to detect whether specific ports are opened on local device. If enabled, Sentinel Firewall Logs gather information about origin of such malicious packets and about ports they try to scan on local device.
          </small>
        </div>
      </div>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-switch"
        >
          <input
            class="custom-control-input"
            id="2"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="2"
          >
            Enable Minipots
          </label>
          <small
            class="form-text text-muted mt-0 mb-3"
          >
            The main purpose of the Sentinel Minipot is to collect authentication information from the login attempts. It is possible to emulate some of the often attacked services - Telnet, HTTP, FTP, and SMTP. The goal is to catch the attacker red-handed when they think they are attacking a real service.
          </small>
        </div>
      </div>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-switch"
        >
          <input
            class="custom-control-input"
            id="3"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="3"
          >
            Enable Survey
          </label>
          <small
            class="form-text text-muted mt-0 mb-3"
          >
            The Turris survey collects information about installed packages, used languages, and operating system versions. Based on this, we can focus on widely used packages and features and provide support.
          </small>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
